src/Workspaces/Remote/Core/AbstractAssetProviderExtensions.AnalyzerReference.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+
+#if NET
+using System.Runtime.Loader;
+#endif
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal static partial class AbstractAssetProviderExtensions
+{
+#if NET
+
+    /// <summary>
+    /// Gate around <see cref="s_checksumToReferenceSet"/> to ensure it is only accessed and updated atomically.
+    /// </summary>
+    private static readonly SemaphoreSlim s_isolatedReferenceSetGate = new(initialCount: 1);
+
+    /// <summary>
+    /// Mapping from checksum for a particular set of assembly references, to the dedicated ALC and actual assembly
+    /// references corresponding to it.  As long as it is alive, we will try to reuse what is in memory.  But once it is
+    /// dropped from memory, we'll clean things up and produce a new one.
+    /// </summary>
+    private static readonly Dictionary<Checksum, WeakReference<IsolatedAssemblyReferenceSet>> s_checksumToReferenceSet = new();
```

Comment:
>s_checksumToReferenceSet [](http://example.com/codeflow?start=94&length=24)

Are checksums ever removed from the Dictionary?

Reply:
they are not.  i'll add a sweep.